lib/matplotlib/tests/test_figure.py

```diff
@@ -861,6 +861,28 @@ def test_hashable_keys(self, fig_test, fig_ref):
         fig_test.subplot_mosaic([[object(), object()]])
         fig_ref.subplot_mosaic([["A", "B"]])
 
+    @pytest.mark.parametrize('str_pattern',
+                             ['abc', 'cab', 'bca', 'cba', 'acb', 'bac'])
+    def test_user_order(self, str_pattern):
+        fig = plt.figure()
+        ax_dict = fig.subplot_mosaic(str_pattern)
+        assert list(str_pattern) == list(ax_dict)
+        assert list(fig.axes) == list(ax_dict.values())
+
+    def test_nested_user_order(self):
+        layout = [
+            ["A", [["B", "C"],
+                   ["D", "E"]]],
+            ["F", "G"],
+            [".", [["H", [["I"],
+                          ["."]]]]]
+        ]
```

Comment:
OK, I now see why this was such a blister.  

Congrats for figuring out how to do this.  OTOH, I don't think getting the nested axes in order with the level above is a necessary complication.  I personally would expect the order to be something like "AFG BCDE H I"  or put another way, I don't think anyone has  a need or the right to expect the nested axes to be in any sort of order with respect to axes at the parent level.  

i.e. in very quick pseudocode:

```
for a in name: 
  if axes:
    alist += [make_axes]
for a in name:
  if nested:
    alist += add_children()
```



We can keep it, and I'm not going to block at all.  But I do think making it *this* clever is making the code less maintainable.  

If you keep this way, I would explain this a *bit* in the comment above so that the reader is expecting considerable complexity.

Reply:
Another super simple way to sort would be to do what you did before, and then assign a numerical value to each axes based on its subplot spec.  If it is down a level, add the parent subplot spec to the child subplot spec divided by 10, etc.  Above you would get

```
A, B,   C,   D,   E,     F, G, H,   I,       J
1, 2.1, 2.2, 2.3, 2.4, 3, 4, 6.1, 7.11, 7.12 
```

Reply:
Visually these two layouts should produce the same figure (to 0th order, they end up with slightly different spacing due to the way nested gridspecs work):

```python

layout1="""
        AABC
        AADE
        FFGG
        FFGG
        ..HI
        ..H.
        """

layout2 = [
            ["A", [["B", "C"],
                   ["D", "E"]]],
            ["F", "G"],
            [".", [["H", [["I"],
                          ["."]]]]]
        ]


fig1, axd1 = plt.subplot_mosaic(layout1)
fig2, axd2 = plt.subplot_mosaic(layout2)
```

I think it would be very surprising if changing how the layout is specified changed the order (now that it has been pointed out to me that people care about the order!).

Reply:
> I think it would be very surprising if changing how the layout is specified changed the order

I think it would be surprising if it didn't change the order!

Reply:
Fair, but that seems much harder to document than "the order is left-to-right and top-down"